File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.java
import java.util.*;
class Solution {
    
    public class go{
        int index;
        int count;
        go(int index,int count){
            this.index=index;
            this.count=count;
        }
    }
    public class first{
        String name;
        int count;
        first(String name,int count){
            this.name=name;
            this.count=count;
        }
    }
    
    public int[] solution(String[] genres, int[] plays) {
        int[] answer = {};
        HashMap<String,Integer> sum=new HashMap<>();
        HashMap<String,List<go>>total=new HashMap<>();
        
        for(int i=0;i<genres.length;i++){
            String gen=genres[i];
            int now=plays[i];
            if(!sum.containsKey(gen)){
                sum.put(gen,0);
                total.put(gen,new ArrayList<>());
            }  
            sum.put(gen,sum.get(gen)+now);
            total.get(gen).add(new go(i,now));
        }
        List<first>test=new LinkedList<>();
        for(String s:sum.keySet()){
            test.add(new first(s,sum.get(s)));
        }
        test.sort((x,y)->{
            return y.count-x.count;
        });
        List<Integer>last=new LinkedList<>();
        for(first f:test){
            List<go>temp=total.get(f.name);
            temp.sort((x,y)->{
                if(x.count!=y.count)
                    return y.count-x.count;
                return x.index-y.index;
            });
            for(int i=0;i<temp.size();i++){
                if(i>1)
                    break;
                last.add(temp.get(i).index);
            }
        }
        answer=new int[last.size()];
        for(int i=0;i<last.size();i++){
            answer[i]=last.get(i);
        }

        return answer;
    }
}

File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.py
def solution(genres, plays):
    answer = []
    
    total_genre=dict()
    total=dict()
    for i in range(len(genres)):
        now_genre=genres[i]
        now_plays=plays[i]
        if now_genre not in total_genre:
            total_genre[now_genre]=plays[i]
            total[now_genre]=[]
        else:
            total_genre[now_genre]+=plays[i]
        total[now_genre].append([plays[i],i])    
    
    #가장 많이 재생된 장르를 수록 하고자 한다
    first=[]
    total_genre=sorted(total_genre.items(),key=lambda x:-x[1])
    for i,_ in total_genre:
        now=total[i]
        now=sorted(now,key=lambda x:(-x[0],x[1]))
        now=now[:2]
        for i1 in now: 
            answer.append(i1[1])
    
    
    return answer